internal/v2/cache/profiles.go

```diff
@@ -0,0 +1,339 @@
+	}
+
+	dr, ok := drs[resourceName]
+	return dr, ok
+}
+
+// CommandExists returns a bool indicating whether the specified command exists for the
+// specified (by name) device. If the specified device doesn't exist, an error is returned.
+func (p *profileCache) CommandExists(profileName string, cmd string, method string) (bool, edgexErr.EdgeX) {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+
+	_, ok := p.deviceProfileMap[profileName]
+	if !ok {
+		errMsg := fmt.Sprintf("failed to find profile %s in cache", profileName)
+		return false, edgexErr.NewCommonEdgeX(edgexErr.KindInvalidId, errMsg, nil)
+	}
+	// Check whether cmd exists in deviceCommands.
+	var deviceCommands map[string][]contract.ResourceOperation
+	if strings.ToLower(method) == common.GetCmdMethod {
+		deviceCommands, _ = p.getResourceOperationsMap[profileName]
+	} else {
+		deviceCommands, _ = p.setResourceOperationsMap[profileName]
+	}
+
+	if _, ok := deviceCommands[cmd]; !ok {
+		errMsg := fmt.Sprintf("failed to find %s command %s in profile %s", method, cmd, profileName)
+		return false, edgexErr.NewCommonEdgeX(edgexErr.KindInvalidId, errMsg, nil)
+	}
+
+	return true, nil
+}
+
+// ResourceOperations returns the ResourceOperations with given command and method.
+func (p *profileCache) ResourceOperations(profileName string, cmd string, method string) ([]contract.ResourceOperation, edgexErr.EdgeX) {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+
+	var ok bool
+	var ros []contract.ResourceOperation
+	var rosMap map[string][]contract.ResourceOperation
+
+	if _, ok = p.deviceProfileMap[profileName]; !ok {
+		errMsg := fmt.Sprintf("failed to find profile %s in cache", profileName)
+		return nil, edgexErr.NewCommonEdgeX(edgexErr.KindInvalidId, errMsg, nil)
+	}
+
+	if strings.ToLower(method) == common.GetCmdMethod {
+		rosMap, ok = p.getResourceOperationsMap[profileName]
+	} else if strings.ToLower(method) == common.SetCmdMethod {
+		rosMap, ok = p.setResourceOperationsMap[profileName]
+	}
+
+	if !ok {
+		errMsg := fmt.Sprintf("failed to find %s ResourceOperations in profile %s", method, profileName)
+		return nil, edgexErr.NewCommonEdgeX(edgexErr.KindInvalidId, errMsg, nil)
+	}
+
+	if ros, ok = rosMap[cmd]; !ok {
+		errMsg := fmt.Sprintf("failed to find %s command in profile %s", cmd, profileName)
+		return nil, edgexErr.NewCommonEdgeX(edgexErr.KindInvalidId, errMsg, nil)
+	}
+
+	return ros, nil
+}
+
+// ResourceOperation returns the first matched ResourceOperation with given deviceResource and method
+func (p *profileCache) ResourceOperation(profileName string, deviceResource string, method string) (contract.ResourceOperation, edgexErr.EdgeX) {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+
+	var ok bool
+	var ro contract.ResourceOperation
+	var rosMap map[string][]contract.ResourceOperation
+
+	if _, ok = p.deviceProfileMap[profileName]; !ok {
+		errMsg := fmt.Sprintf("failed to find profile %s in cache", profileName)
+		return ro, edgexErr.NewCommonEdgeX(edgexErr.KindInvalidId, errMsg, nil)
+	}
```

Comment:
Should be a helper func since used many times.

Somethin like this
```
func (p *profileCache) verifyProfileExists(name string)  errors.EdgeX {
    if _, ok = p.deviceProfileMap[profileName]; !ok {
          errMsg := fmt.Sprintf("failed to find profile %s in cache", profileName)
		return edgexErr.NewCommonEdgeX(edgexErr.KindInvalidId, errMsg, nil)
    }
    return nil
}
```

Call then changes to

```
if err := p.verifyProfileExists(profileName); err != nil {
    return ro, err
}
```